src/main/scala/com/databricks/sparkdl/ImageUtils.scala

```diff
@@ -42,74 +44,127 @@ private[sparkdl] object ImageUtils {
           | image of size ($height, $width, $channels).
        """.stripMargin
     )
-    val image = new BufferedImage(width, height, BufferedImage.TYPE_3BYTE_BGR)
 
-    var offset, h = 0
-    var r, g, b: Byte = 0
-    while (h < height) {
-      var w = 0
-      while (w < width) {
-        channels match {
-          case 1 =>
-            b = imageData(offset)
-            g = b
-            r = b
-          case 3 =>
-            b = imageData(offset)
-            g = imageData(offset + 1)
-            r = imageData(offset + 2)
-          case _ =>
-            require(false, s"`Channels` must be 1 or 3, got $channels.")
-        }
+    val image = channels match {
+      case 1 => new BufferedImage(width, height, BufferedImage.TYPE_BYTE_GRAY)
+      case 3 => new BufferedImage(width, height, BufferedImage.TYPE_3BYTE_BGR)
+      case 4 => new BufferedImage(width, height, BufferedImage.TYPE_4BYTE_ABGR)
+    }
 
-        val color = new Color(r & 0xff, g & 0xff, b & 0xff)
-        image.setRGB(w, h, color.getRGB)
-        offset += channels
-        w += 1
+    val isGray = image.getColorModel.getColorSpace.getType == ColorSpace.TYPE_GRAY
+    val hasAlpha = image.getColorModel.hasAlpha
+    var offset = 0
+    // Grayscale images in Java require special handling to get the correct intensity
```

Comment:
The logic here is basically the inverse of Spark's logic for reading Grayscale images into a data array (this code sets the pixels of a grayscale image from a data array): https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/image/ImageSchema.scala#L159

For context, if we set grayscale image data using `image.setRGB(w, h, rgbInt)`, calling `image.getRGB(w, h)` won't necessarily return the same result (see [this gist](https://gist.github.com/smurching/d6e918a84c79155b616f9339b30fdfca)). Hence we set grayscale image data directly in the raster here.